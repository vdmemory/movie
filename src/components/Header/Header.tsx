import React, { useEffect } from 'react'

import styles from "./Header.module.scss";
import { useDispatch, useSelector } from 'react-redux';
import { selectSearchMovies, selectDetailsMovie } from '../../store/selectors';
import { getSearchMoviesRequest, getDetailsMovieRequest } from '../../store/actions';

const Header: React.FC = () => {
    const dispatch = useDispatch();

    const {searchMoviesData, searchMoviesLoading, searchMoviesError} = useSelector(selectSearchMovies);

    const {detailsMovieData} = useSelector(selectDetailsMovie)

    console.log(searchMoviesData);

    useEffect(() => {
        dispatch(getSearchMoviesRequest('hgvfghvfytfytvytyt'))
        dispatch(getDetailsMovieRequest("tt0093692"));
    }, [dispatch])

    return <div className={styles.header}>Header</div>;
};

export default Header;
