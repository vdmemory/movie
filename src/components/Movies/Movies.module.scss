.main {
    padding: 15px;
    width: 100%;
    max-width: 1200px;
    margin: 0 auto;

    & .movies {
        display: grid;
        grid-gap: 1em 1em;
        grid-template-columns: repeat(auto-fill, minmax(9em, 1fr));
        margin: 0;
        padding: 0;
    }
}

.message {
    text-align: center;
}

