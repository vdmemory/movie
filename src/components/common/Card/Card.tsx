import React from "react";

import noposter from "./noposter.png"
import styles from "./Card.module.scss";

type Props = {
    title: string;
    poster: string;
    year: string;
    id: string;
    onClick(id: string): void;
};

const Card: React.FC<Props> = ({ title, poster, onClick, id, year }) => {
    const handleClick = () => {
        onClick(id);
    };

    return (
        <div className={styles.movie}>
            <figure onClick={handleClick}>
                <img alt={title} src={poster !== "N/A" ? poster : noposter} />
                <h2>{`${title} (${year})`}</h2>
            </figure>
        </div>
    );
};

export default Card;
