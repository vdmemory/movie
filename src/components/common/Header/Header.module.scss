.header {
    padding: 15px;
    background: aliceblue;

    & .title {
        font-family: 'Roboto';
        font-size: 22px;
        text-align: center;
    }
}





