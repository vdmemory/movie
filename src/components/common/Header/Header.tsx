import React, { Dispatch } from "react";
import Search from "../Search/Search";

import styles from "./Header.module.scss";

type Props = {
    placeholder: string;
    titleHeader: string;
    titleSearch: string;
    type: string;
    dispatch: Dispatch<any>;
};

const Header: React.FC<Props> = ({
    titleHeader,
    titleSearch,
    type,
    placeholder,
    dispatch,
}) => {
    return (
        <header className={styles.header}>
            <h1 className={styles.title}>{titleHeader}</h1>
            <Search
                placeholder={placeholder}
                title={titleSearch}
                type={type}
                dispatch={dispatch}
            />
        </header>
    );
};

export default Header;
