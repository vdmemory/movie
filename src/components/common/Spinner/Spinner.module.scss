.wrapSpinner {
    width: 100%;
    height: 20vh;
    display: flex;
    justify-content: center;
    align-items: center;

    & .spinner {
        width: 2rem;
        height: 2rem;
        border: 3px solid transparent;
        border-top-color: rgba(black, .5);
        border-bottom-color: rgba(black, .5);
        border-radius: 50%;
        animation: rotation .8s ease infinite;
    }
}

@keyframes rotation {
    from {
        transform: rotate(0deg);
    }

    to {
        transform: rotate(360deg);
    }
}