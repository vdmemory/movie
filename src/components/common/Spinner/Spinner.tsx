import React from "react";

import styles from "./Spinner.module.scss";

const Spinner: React.FC = () => (
    <div className={styles.wrapSpinner}>
        <div className={styles.spinner}></div>
    </div>
); 
export default Spinner;
