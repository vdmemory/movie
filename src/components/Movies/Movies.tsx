import React from "react";

import { SearchMovies } from "../../api/searchMovies";
import Header from "../common/Header/Header";
import Card from "../common/Card/Card";
import Spinner from "../common/Spinner/Spinner";
import { useCustomHistory } from "../../hooks";
import { useMovies } from "./hooks";
import styles from "./Movies.module.scss";

const dataHeader = {
    titleHeader: "Find Movies, TV shows and more ...",
    titleSearch: "Search",
    type: "text",
    placeholder: "search movie ...",
};

const Movies: React.FC = () => {
    const { handlePushHistory } = useCustomHistory();
    const {
        searchMoviesData,
        searchMoviesLoading,
        searchMoviesError,
    } = useMovies();

    const renderMovies = (searchMovies: SearchMovies) => {
        const { imdbId } = searchMovies;

        return (
            <Card
                key={`movie${imdbId}`}
                id={imdbId}
                onClick={() => handlePushHistory(`movie/${imdbId}`)}
                {...searchMovies}
            />
        );
    };

    const renderMain = () => {
        if (searchMoviesLoading) return <Spinner />;

        if (
            searchMoviesData &&
            searchMoviesData.length === 0 &&
            !searchMoviesError
        )
            return <p className={styles.message}>No results ...</p>;

        if (searchMoviesError)
            return <p className={styles.message}>{searchMoviesError}</p>;

        return (
            <main className={styles.main}>
                <div className={styles.movies}>
                    {searchMoviesData.map(renderMovies)}
                </div>
            </main>
        );
    };

    return (
        <>
            <Header {...dataHeader} />
            {renderMain()}
        </>
    );
};

export default Movies;
