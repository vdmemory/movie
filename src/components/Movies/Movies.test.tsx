import React from "react";
import { createRenderer } from "react-test-renderer/shallow";
import { useSelector } from "react-redux";
import ReactPaginate from "react-paginate";

import Movies from "./Movies";
import Spinner from "../common/Spinner/Spinner";
import Header from "../common/Header/Header";
import Card from "../common/Card/Card";
import { initialState } from "../../store/reducer";
import { SearchMovies } from ".././../api/searchMovies";

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useHistory: () => ({
        push: jest.fn(),
    }),
}));

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: () => mockDispatch,
}));

const setup = () => {
    const renderer = createRenderer();
    renderer.render(<Movies />);
    const output = renderer.getRenderOutput();
    return { output, renderer };
};

const stateLoading = {
    ...initialState,
    loading: {
        ...initialState.loading,
        searchMovies: true,
    },
};

const stateError = {
    ...initialState,
    error: {
        ...initialState.error,
        searchMovies: "Error message",
    },
};

const stateDataСontent = {
    ...initialState,
    searchMovies: {
        totalResults: "100",
        search: [
            {
                title: "this title",
                year: "this year",
                imdbId: "this imdbId",
                type: "this type",
                poster: "this poster",
            },
        ] as SearchMovies[],
    },
    searchValue: "test",
};

describe("searchMovies Component", () => {
    describe("Renderer of the component and its components, if no search query has been entered", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(initialState);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render Header component and check class", () => {
            const { output } = setup();
            const [component] = output.props.children;
            expect(component.type).toBe(Header);
        });

        test("Should render no results message if search.length === 0 and check class", () => {
            const { output } = setup();
            const [, error] = output.props.children;
            expect(error.type).toBe("p");
            expect(error.props.className).toBe("message");
            expect(error.props.children).toBe("No results ...");
        });
    });

    describe("Renderer of the component and its components, if the loading flag is true", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(stateLoading);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render Spinner component and check class", () => {
            const { output } = setup();
            const [, spinner] = output.props.children;
            expect(spinner.type).toBe(Spinner);
        });
    });

    describe("Renderer of the component and its components, if the error message", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(stateError);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render Error message and check class", () => {
            const { output } = setup();
            const [, error] = output.props.children;
            expect(error.type).toBe("p");
            expect(error.props.className).toBe("message");
            expect(error.props.children).toBe("Error message");
        });
    });

    describe("Renderer of the component and its components, if a search query was entered", () => {
        beforeEach(() => {
            // @ts-expect-error
            useSelector.mockImplementation((callback) => {
                return callback(stateDataСontent);
            });
        });
        afterEach(() => {
            // @ts-expect-error
            useSelector.mockClear();
        });
        test("Should renders correctly", () => {
            const { output } = setup();
            expect(output).toMatchSnapshot();
        });

        test("Should render Main container and check class", () => {
            const { output } = setup();
            const [, main] = output.props.children;
            expect(main.type).toBe("main");
            expect(main.props.className).toBe("main");
            expect(main.props.children.props.className).toBe("movies");
        });

        test("Should render Card component", () => {
            const { output } = setup();
            const [, main] = output.props.children;
            const card = main.props.children.props.children[0];
            expect(card.type).toBe(Card);
        });

        test("Should render Pagination component and check class", () => {
            const { output } = setup();
            const [, , paginate] = output.props.children;
            expect(paginate.type).toBe(ReactPaginate);
            expect(paginate.props.containerClassName).toBe("pagination");
        });

        test("Should call handleChangePaginate on click", () => {
            const { output } = setup();
            const [, , paginate] = output.props.children;
            paginate.props.onPageChange(2);

            expect(mockDispatch).toBeCalled();
        });
    });
});
