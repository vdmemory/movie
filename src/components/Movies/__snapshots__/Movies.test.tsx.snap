// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchMovies Component Renderer of the component and its components, if a search query was entered Should renders correctly 1`] = `
<React.Fragment>
  <Header
    dispatch={[MockFunction]}
    placeholder="search movie ..."
    titleHeader="Find Movies, TV shows and more ..."
    titleSearch="Search"
    type="text"
  />
  <main
    className="main"
  >
    <div
      className="movies"
    >
      <Card
        id="this imdbId"
        onClick={[Function]}
        poster="this poster"
        title="this title"
        type="this type"
        year="this year"
      />
    </div>
  </main>
  <o
    activeClassName="active"
    breakClassName="break-me"
    breakLabel="..."
    containerClassName="pagination"
    disableInitialCallback={false}
    disabledClassName="disabled"
    initialPage={0}
    marginPagesDisplayed={2}
    nextAriaLabel="Next page"
    nextClassName="next"
    nextLabel="next"
    onPageChange={[Function]}
    pageCount={10}
    pageRangeDisplayed={6}
    previousAriaLabel="Previous page"
    previousClassName="previous"
    previousLabel="previous"
  />
</React.Fragment>
`;

exports[`searchMovies Component Renderer of the component and its components, if no search query has been entered Should renders correctly 1`] = `
<React.Fragment>
  <Header
    dispatch={[MockFunction]}
    placeholder="search movie ..."
    titleHeader="Find Movies, TV shows and more ..."
    titleSearch="Search"
    type="text"
  />
  <p
    className="message"
  >
    No results ...
  </p>
</React.Fragment>
`;

exports[`searchMovies Component Renderer of the component and its components, if the error message Should renders correctly 1`] = `
<React.Fragment>
  <Header
    dispatch={[MockFunction]}
    placeholder="search movie ..."
    titleHeader="Find Movies, TV shows and more ..."
    titleSearch="Search"
    type="text"
  />
  <p
    className="message"
  >
    Error message
  </p>
</React.Fragment>
`;

exports[`searchMovies Component Renderer of the component and its components, if the loading flag is true Should renders correctly 1`] = `
<React.Fragment>
  <Header
    dispatch={[MockFunction]}
    placeholder="search movie ..."
    titleHeader="Find Movies, TV shows and more ..."
    titleSearch="Search"
    type="text"
  />
  <Spinner />
</React.Fragment>
`;
